Fix useZoomOut inserter behavior

jeryj opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/66328

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Track if the useZoomOut hook changed the zoom level for the user. If we never changed it, don't change the zoom level on unmount.

Request from @richtabor: If we did change zoom level via the `useZoomOut` hook (controlled mode) _always `resetZoomLevel()` on unmount_.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Fixes a bug where zoom out was engaged/disengaged incorrectly on unmount. 

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Adds a ` controlZoomLevelRef`  that tracks if we ever changed the zoom level for the user. When unmounting, if we changed the zoom level, reset it _unless they manually toggled the zoom level after us_. In that condition, `controlZoomLevelRef` is set to `false`.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

The key concepts here are:
- If we most recently changed the zoom level for them (in or out), we always resetZoomLevel() level when unmounting.
- If the user most recently changed the zoom level (manually toggling), we do nothing when unmounting.
 
So, with these in mind, we have two scenarios to watch for:
- Closing the inserter always resets the zoom level if the inserter changed the zoom level (clicking on blocks tab from zoomed out state OR clicking on patterns or media tab from zoomed in state)
- Closing the inserter NEVER affects zoom level if zoom state was manually toggled via the zoom out button. However, if you manually toggle, then click to a tab that sets the zoom state, we are now controlling it again.


1. Enter Zoom out manually
2. Open the inserter.
3. The pattern tab should be selected.
4. Exit zoom out manually.
5. Close the inserter.
6. Editor should remain in non-zoomed state (regular editing)

From full screen mode
1. Open inserter
2. Click patterns tab
3. Zoom out should engage
4. Close inserter
5. Zoom out should disengage

From full screen mode, exit zoom out manually
1. Open inserter
2. Click patterns tab
3. Zoom out should engage
4. Manually exit zoom out
5. Close inserter
6. Zoom out should not engage (remain in full screen mode)

## Screen Recording

https://github.com/user-attachments/assets/082ce511-c038-4de4-aa92-5c35f7ef6a5f

Mamaduka committed:
Block Editor: Fix 'useZoomOut' hook conflicts

jeryj committed:
Add controlled concept to useZoomOut hook

jeryj committed:
Add e2e tests to cover scenarios for useZoomOut hook exiting

jeryj committed:
Codify useZoomOut behaviors

jeryj committed:
Fix bug where mounting to controlled state would get rest to not controlled state

Mamaduka commented on packages/block-editor/src/hooks/use-zoom-out.js:
```diff
@@ -12,32 +12,53 @@ import { unlock } from '../lock-unlock';
 	);
 	const { isZoomOut } = unlock( useSelect( blockEditorStore ) );
 
+	const isZoomedOut = isZoomOut();
```
Calling a selector during the render isn't recommended.

jeryj replied:
Thanks for catching! Updated to call it within a `useSelect`.

jeryj committed:
remove unnecessary pattern button click in test

jeryj committed:
Refactor if statements for setting controlZoomLevelRef

jeryj committed:
Refactor useZoomOut for readability

jeryj committed:
Refactor to get isZoomedOut within useSelect

richtabor commented:
Experience wise, this feels nice. It stays zoomed out when expected to, and zooms back in when appropriate. 

https://github.com/user-attachments/assets/a4ef232c-4b07-47e3-8b4e-cdf81565a567

richtabor reviewed (approved):
This feels nice! You don't end up in unexpected scenarios where zoom out is auto re-engaged.

jeryj committed:
Big refactor of zoom out insterter tests

Mamaduka commented on test/e2e/specs/site-editor/site-editor-inserter.spec.js:
```diff
@@ -64,4 +62,396 @@ test.describe( 'Site Editor Inserter', () => {
+		const blockLibrary = InserterUtils.getBlockLibrary();
+
+		await zoomOutButton.click();
+		await expect( await InserterUtils.getZoomCanvas() ).toBeVisible();
```
```suggestion
		await expect( InserterUtils.getZoomCanvas() ).toBeVisible();
```

I think there's no need to `await` canvas locator. Usually, you would write it like this - `await expect( page.locator( '.is-zoomed-out' ) ).toBeVisible();`.

Linter will remove extra await if you try to put it there.

Mamaduka commented on test/e2e/specs/site-editor/site-editor-inserter.spec.js:
```diff
@@ -64,4 +62,396 @@ test.describe( 'Site Editor Inserter', () => {
+	} );
 } );
+
+class InserterUtils {
```
IMO, the utils class here seems redundant. It just holds aliases for simple locator calls. Why not just inline them?

jeryj replied:
I redid them today to make them a lot more readable. They were pretty scattered and hard to piece together IMO since there are so many conditions to test.

I'm still using the `InserterUtils` and also added a `ZoomUtils`. They do hold basic locators and some `expects`, but I think it make it way more readable. I was able to remove a lot of comments and just have the method names document the actions.

Mamaduka replied:
Makes sense!

Mamaduka reviewed (approved):
Thank you, @jeryj!

The fix looks good to me ✅ I just left a couple of notes regarding e2e tests.

P.S. Rebasing should fix the failing e2e test.

jeryj committed:
Refactor to use test steps

jeryj committed:
Refactor from toggle to open/close methods for readability

jeryj committed:
use aria-selected instead of data-active-item to check for active block tab

jeryj committed:
Remove unused seleectors

jeryj committed:
Add deleteAllTemplates calls back in

jeryj committed:
Normalize preload test by resetting preferences before the test starts

This test kept failing for me, and I isolated it to if a test before it had run await admin.visitSiteEditor(), then the preload test would fail because the user had previously visited the site editor. By resetting preferences before the test runs, it allows a cleaner enviroment and the test has consistent results.

jeryj commented on test/e2e/specs/site-editor/preload.spec.js:
```diff
@@ -6,6 +6,7 @@ const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
 test.describe( 'Preload', () => {
 	test.beforeAll( async ( { requestUtils } ) => {
 		await requestUtils.activateTheme( 'emptytheme' );
+		await requestUtils.resetPreferences();
```
This should really be in a different PR, but I wanted to see if adding this would fix the e2e issue I'm running into. [More context in the PR that implemented the Preload test](https://github.com/WordPress/gutenberg/pull/67497#issuecomment-2523847534).

tl;dr: The Preload test will have different results depending on what runs before it, and this should help normalize it. I just happened to run into the issue by having my test run on the same shard.

jeryj replied:
It did work -- I would have liked this fix a different PR but it auto-merged once the tests passed.
